### PR TITLE
Generate {Java, Type}Script clients.

### PR DESCRIPTION
**Pull request by nhynes**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,8 +1,8 @@
 [package]
 name = "oasis-cli"
-version = "0.1.7"
+version = "0.2.0"
```
Comment (ennsharma) [inline Cargo.toml]: No patch versions, right?

File: src/workspace.rs [hunk 2]
```diff
@@ -82,62 +83,104 @@ impl Workspace {
...
+        top_deps.sort_by_key(|deps| -(deps.len() as isize));
+        for dep_targets in top_deps.into_iter() {
+            for dep_target in dep_targets.into_iter() {
+                if !build_plan.iter().any(|t| t.name == dep_target.name) {
+                    build_plan.push(dep_target);
+                }
+            }
+        }
+
+        // Add the top targets to tbe build plan.
```
Comment (ennsharma) [inline src/workspace.rs]:
```suggestion
        // Add the top targets to the build plan.
```

File: src/workspace.rs [hunk 3]
```diff
@@ -171,146 +214,211 @@ impl Workspace {
...
+                manifest_path: PathBuf::from(pkg.manifest_path),
+                kind: ProjectKind::Rust,
+                targets: Vec::new(),
+            });
+            let proj_ref = unsafe { &*(&*proj as *const Project) }; // @see `struct Workspace`
+            for target in pkg.targets {
+                let is_buildable = target.kind[0] == "bin"; // may include unit tests
+                let is_testable = target.kind[0] == "test"; // integration tests
+
+                let mut phases = Phases::CLEAN /* Cargo projects are always cleanable */;
```
Comment (ennsharma) [inline src/workspace.rs]: Why the two different comment styles?

File: src/gen/typescript.rs [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+
+        #(#imports)*
+
+        #(#type_defs)*
+
+        export class #service_ident {
+            public static BYTECODE_URL = #bytecode_url_str;
+
+            private constructor(readonly address: oasis.Address, private gateway: oasis.Gateway) {}
+
```
Comment (eauge) [inline src/gen/typescript.rs]: do we need a create method as well? Or is there a custom create method for each service?

Comment (nhynes) [reply src/gen/typescript.rs]: yes, that's `deploy`, which is `#deploy_function`

**Pull request merged**